### PR TITLE
Add a SubArea sample app with tests

### LINKED ISSUE #754: Add tests for route data constraints in url-generation/matching
This would be something like subareas. We want a functional test to make sure the scenario stays working. Needs to include conventional routing and attribute routing.

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
Fixes #754

### CODE REVIEW

File: samples/MvcSubAreaSample.Web/Areas/Restaurant/Areas/Hours/Views/Home/Index.cshtml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+﻿@*
+    For more information on enabling MVC for empty projects, visit http://go.microsoft.com/fwlink/?LinkID=397860
+*@
```
Comment (rynowak) [inline samples/MvcSubAreaSample.Web/Areas/Restaurant/Areas/Hours/Views/Home/Index.cshtml]: Delete

Comment (rynowak) [inline samples/MvcSubAreaSample.Web/Areas/Restaurant/Areas/Hours/Views/Home/Index.cshtml]: all of em

File: samples/MvcSubAreaSample.Web/Startup.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        {
+            services.Configure<RazorViewEngineOptions>(options =>
+            {
+                options.ViewLocationExpanders.Add(new SubAreaViewLocationExpander());
+            });
+
+            services.AddMvc(options =>
+            {
+                //options.Filters.Add(typeof(PassThroughAttribute), order: 17);
+                options.Filters.Add(new FormatFilterAttribute());
```
Comment (rynowak) [inline samples/MvcSubAreaSample.Web/Startup.cs]: Is this used?

File: samples/MvcSubAreaSample.Web/Startup.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        {
+            // Add MVC to the request pipeline.
+            app.UseMvc(routes =>
+            {
+                routes.MapRoute("areaBase", "{area:exists}",
+                    new
+                    {
+                        controller = "Home",
+                        action = "Index"
+                    });
```
Comment (rynowak) [inline samples/MvcSubAreaSample.Web/Startup.cs]: What's this needed for? IE why can't defaults express this?

File: samples/MvcSubAreaSample.Web/Startup.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                    "controllerActionRoute",
+                    "{controller}/{action}",
+                    new { controller = "Home", action = "Index" },
+                    constraints: null,
+                    dataTokens: new { NameSpace = "default" });
+
+                routes.MapRoute(
+                    "controllerRoute",
+                    "{controller}",
+                    new { controller = "Home" });
```
Comment (rynowak) [inline samples/MvcSubAreaSample.Web/Startup.cs]: ???

File: samples/MvcSubAreaSample.Web/SubAreaAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+{
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = false, Inherited = true)]
+    public class SubAreaAttribute : RouteConstraintAttribute
+    {
+        public SubAreaAttribute(string name)
+            : base("subarea", name, blockNonAttributedActions: true)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentException("Area name must not be empty", nameof(name));
```
Comment (rynowak) [inline samples/MvcSubAreaSample.Web/SubAreaAttribute.cs]: subarea

File: samples/MvcSubAreaSample.Web/SubAreaViewLocationExpander.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+using System.Linq;
+using Microsoft.AspNet.Mvc.Razor;
+
+namespace MvcSubAreaSample.Web
+{
+    public class SubAreaViewLocationExpander : IViewLocationExpander
+    {
+        private const string _subAreaKey = "subarea";
+
+        public IEnumerable<string> ExpandViewLocations(ViewLocationExpanderContext context,
```
Comment (rynowak) [inline samples/MvcSubAreaSample.Web/SubAreaViewLocationExpander.cs]: #YOLOBREAK

File: samples/MvcSubAreaSample.Web/SubAreaViewLocationExpander.cs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+            IEnumerable<string> viewLocations)
+        {
+            if (context.Values.ContainsKey(_subAreaKey))
+            {
+                var subArea = context.Values[_subAreaKey];
```
Comment (rynowak) [inline samples/MvcSubAreaSample.Web/SubAreaViewLocationExpander.cs]: /cc @pranavkm - do we need to do normalization (casing) here?

File: samples/MvcSubAreaSample.Web/Views/Shared/_Layout.cshtml [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+                    <span class="icon-bar"></span>
+                    <span class="icon-bar"></span>
+                </button>
+                <a asp-controller="Home" asp-action="Index" class="navbar-brand">MvcSubAreaSample.Web</a>
+            </div>
+            <div class="navbar-collapse collapse">
+                <ul class="nav navbar-nav">
+                    <li><a asp-controller="Home" asp-action="Index">Home</a></li>
+                    <li><a asp-controller="Home" asp-action="About">About</a></li>
+                    <li><a asp-controller="Home" asp-action="Contact">Contact</a></li>
```
Comment (rynowak) [inline samples/MvcSubAreaSample.Web/Views/Shared/_Layout.cshtml]: I'd like to see some link generation used here to the various area/subareas

**Comment by rynowak** [comment]
:watch:

File: samples/MvcSubAreaSample.Web/SubAreaViewLocationExpander.cs (hunk 7 above)
Comment (pranavkm) [inline samples/MvcSubAreaSample.Web/SubAreaViewLocationExpander.cs]: Yeah - Use `RazorViewEngine.GetNormalizedRouteValue` to get the value.

File: samples/MvcSandbox/wwwroot/_references.js [hunk 9]
```diff
@@ -1,6 +1,4 @@
 ﻿/// <autosync enabled="true" />
-/// <reference path="../gulpfile.js" />
```
Comment (pranavkm) [inline samples/MvcSandbox/wwwroot/_references.js]: Undo?

File: samples/MvcSubAreaSample.Web/Areas/Restaurant/Areas/Hours/Controllers/HomeController.cs [hunk 10]
```diff
@@ -0,0 +1,18 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNet.Mvc;
+
+namespace MvcSubAreaSample.Web.Areas.Restaurant.Hours.Controllers
```
Comment (pranavkm) [inline samples/MvcSubAreaSample.Web/Areas/Restaurant/Areas/Hours/Controllers/HomeController.cs]: `namespace MvcSubAreaSample`? Just so that people don't assume the namespace has to do anything with the subarea discovery.

File: samples/MvcSubAreaSample.Web/Startup.cs [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+            services.AddMvc(options =>
+            {
+                //options.Filters.Add(typeof(PassThroughAttribute), order: 17);
+                options.Filters.Add(new FormatFilterAttribute());
+            });
+
+            return services.BuildServiceProvider();
+        }
+
+        public IConfigurationRoot Configuration { get; set; }
```
Comment (pranavkm) [inline samples/MvcSubAreaSample.Web/Startup.cs]: Remove

File: samples/MvcSubAreaSample.Web/Startup.cs [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+                        action = "Index"
+                    });
+                routes.MapRoute("subarearoute", "{area:exists}/{subarea:exists}/{controller}/{action}");
+                routes.MapRoute("areaRoute", "{area:exists}/{controller}/{action}");
+                routes.MapRoute(
+                    "controllerActionRoute",
+                    "{controller}/{action}",
+                    new { controller = "Home", action = "Index" },
+                    constraints: null,
+                    dataTokens: new { NameSpace = "default" });
```
Comment (pranavkm) [inline samples/MvcSubAreaSample.Web/Startup.cs]: What's `dataTokens: new { NameSpace = "default" }` used for?

File: samples/MvcSubAreaSample.Web/SubAreaViewLocationExpander.cs (hunk 7 above)
Comment (pranavkm) [inline samples/MvcSubAreaSample.Web/SubAreaViewLocationExpander.cs]: I guess this would be fine (since it's simply pulling the value out of the ViewLocationExpander dictionary. `PopulateValues` would need to change to use the GetNormalizedRouteValue method.

File: samples/MvcSubAreaSample.Web/Views/Home/Index.cshtml [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+    ViewData["Title"] = "Home Page";
+    Layout = "/Views/Shared/_Layout.cshtml";
+}
+
+<div class="row">
+    <div class="col-md-3">
+        <h2>Areas</h2>
+        <ul>
+            <li><a href="/Restaurant">Restaurant</a></li>
+            </ul>
```
Comment (pranavkm) [inline samples/MvcSubAreaSample.Web/Views/Home/Index.cshtml]: Weird indentation

File: samples/MvcSubAreaSample.Web/Views/Shared/_Layout.cshtml [hunk 14]
```diff
@@ -0,0 +1,79 @@
+﻿<!DOCTYPE html>
+<html>
+<head>
+    <meta charset="utf-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>@ViewData["Title"] - MvcSubAreaSample.Web</title>
+
+    <environment names="Development">
```
Comment (pranavkm) [inline samples/MvcSubAreaSample.Web/Views/Shared/_Layout.cshtml]: Could we simplify this? We don't need to demonstrate tag helpers here.

File: samples/MvcSubAreaSample.Web/Areas/Restaurant/Areas/Hours/Controllers/HomeController.cs (hunk 10 above)
Comment (ryanbrandenburg) [inline samples/MvcSubAreaSample.Web/Areas/Restaurant/Areas/Hours/Controllers/HomeController.cs]: I'm torn between like you said, avoiding confusion and the fact that putting these in at the base namespace is never how someone would/should do it in a real site. @rynowak thoughts?

Comment (rynowak) [inline samples/MvcSubAreaSample.Web/Areas/Restaurant/Areas/Hours/Controllers/HomeController.cs]: My philosophy is to just write code in a sample the way a user would write it. Everything you see in a sample that you don't immediately grok has a mental cost associated.

I tend to avoid using things like `.Controllers` in namespaces in general. Using the area/subarea in the namespace seems pretty usual, since areas are really about code organization.

Comment (ryanbrandenburg) [inline samples/MvcSubAreaSample.Web/Areas/Restaurant/Areas/Hours/Controllers/HomeController.cs]: Ah, so @pranavkm was right either way and I just understand recommended namespace usage less than I thought. Fair deuce, will fix.

File: samples/MvcSubAreaSample.Web/Startup.cs (hunk 12 above)
Comment (ryanbrandenburg) [inline samples/MvcSubAreaSample.Web/Startup.cs]: It's just something I missed in the cleanup of the project I used as a base. I'm removing the whole route since it's redundant with the defaults.

File: samples/MvcSubAreaSample.Web/Startup.cs (hunk 3 above)
Comment (ryanbrandenburg) [inline samples/MvcSubAreaSample.Web/Startup.cs]: Without this route "/Restaurant" gives a 404. Is there functionality I'm missing which would allow you to set general defaults (IE controller=Home and Action=Index) for all routes, or am I misunderstanding the question?

Comment (rynowak) [inline samples/MvcSubAreaSample.Web/Startup.cs]: Add defaults to the route above.
```
routes.MapRoute("subarearoute", "{area:exists}/{subarea:exists}/{controller=Home}/{action=Index}");
```

**Comment by ryanbrandenburg** [comment]
:up: :date:

File: samples/MvcSubAreaSample.Web/Areas/Restaurant/Views/Home/Index.cshtml [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+    Layout = "/Views/Shared/_Layout.cshtml";
+    ViewData["Title"] = "Home Page";
+}
+
+<div class="row">
+    <div class="col-md-3">
+        <h2>Areas</h2>
+        <ul>
+            <li><a href="/Restaurant/Menu">Menu</a></li>
+            <li><a href="/Restaurant/Hours">Hours</a></li>
```
Comment (rynowak) [inline samples/MvcSubAreaSample.Web/Areas/Restaurant/Views/Home/Index.cshtml]: LOLWUT? Use `@Html.ActionLink`

File: samples/MvcSubAreaSample.Web/Startup.cs (hunk 3 above)
Comment (rynowak) [inline samples/MvcSubAreaSample.Web/Startup.cs]: Again, why is this here?

File: samples/MvcSubAreaSample.Web/Views/Home/Index.cshtml [hunk 16]
```diff
@@ -0,0 +1,13 @@
+﻿@{
+    ViewData["Title"] = "Home Page";
+    Layout = "/Views/Shared/_Layout.cshtml";
+}
+
+<div class="row">
+    <div class="col-md-3">
+        <h2>Areas</h2>
+        <ul>
+            <li><a href="/Restaurant">Restaurant</a></li>
```
Comment (rynowak) [inline samples/MvcSubAreaSample.Web/Views/Home/Index.cshtml]: `@Html.ActionLink`

File: samples/MvcSubAreaSample.Web/Views/Shared/_Layout.cshtml [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+                    <span class="icon-bar"></span>
+                    <span class="icon-bar"></span>
+                </button>
+                <a asp-controller="Home" asp-action="Index" class="navbar-brand">MvcSubAreaSample.Web</a>
+            </div>
+            <div class="navbar-collapse collapse">
+                <ul class="nav navbar-nav">
+                    <li><a asp-controller="Home" asp-action="Index">Home</a></li>
+                    <li><a asp-route-area="Restaurant" asp-route-subarea="Menu" asp-controller="Home" asp-action="Index">Menu</a></li>
+                    <li><a asp-route-area="Restaurant" asp-route-subarea="Hours" asp-controller="Home" asp-action="Index">Hours</a></li>
```
Comment (rynowak) [inline samples/MvcSubAreaSample.Web/Views/Shared/_Layout.cshtml]: :heart_eyes_cat:

**Comment by rynowak** [comment]
:watch:

File: samples/MvcSubAreaSample.Web/Startup.cs (hunk 3 above)
Comment (ryanbrandenburg) [inline samples/MvcSubAreaSample.Web/Startup.cs]: I've condensed this down more since you asked, but for some reason Github is still showing an old version instead of marking the comment as outdated. Please look at this one under "Files Changed" to see how it's actually looking right now.

File: samples/MvcSubAreaSample.Web/Views/Home/Index.cshtml (hunk 16 above)
Comment (ryanbrandenburg) [inline samples/MvcSubAreaSample.Web/Views/Home/Index.cshtml]: I ended up doing the taghelper style of these instead since it looks a bit cleaner to me and I was already using them elsewhere, so mixing the two seemed weird. If there's a reason this specifically should be @Html.Actionlink let me know.

**Comment by ryanbrandenburg** [comment]
:up: :date:

File: samples/MvcSubAreaSample.Web/Startup.cs [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+                routes.MapRoute("areaRoute", "{area:exists}/{controller}/{action}",
+                    new
+                    {
+                        controller = "Home",
+                        action = "Index"
+                    });
+                routes.MapRoute(
+                    "controllerActionRoute",
+                    "{controller}/{action}",
+                    new { controller = "Home", action = "Index" });
```
Comment (rynowak) [inline samples/MvcSubAreaSample.Web/Startup.cs]: If you want to be super cool you can use inline defaults. `"{controller=Home}/{action=Index}"`

**Comment by rynowak** [comment]
:shipit:

File: samples/MvcSubAreaSample.Web/Startup.cs (hunk 18 above)
Comment (ryanbrandenburg) [inline samples/MvcSubAreaSample.Web/Startup.cs]: I DO want to be that cool, that's a new one for me and it looks cleaner, thanks!
